docs/SAW.rst

```diff
@@ -2,9 +2,419 @@
+
+The tag field is named ``type``. This tag value can be:
+
+``primitive type``
+  An LLVM primitive type. This is an additional field ``primitive`` which can be any of the following:
+
+  - ``label``: An LLVM label.
+  - ``void``: The LLVM void type.
+  - ``integer``: An LLVM integer. There is an additional field ``size``, an integer giving the number of
+    bytes in the integer type.
+  - ``float``: An LLVM float. There is an additional field ``float type`` which can be any of the following:
+
+    + ``half``
+    + ``float``
+    + ``double``
+    + ``fp128``
+    + ``x86_fp80``
+    + ``PPC_fp128``
+
+  - ``X86mmx``: An x86 SIMD instruction.
+  - ``metadata``: LLVM metadata.
+
+``type alias``
+  A type alias. There is an additional field ``alias of``, which identifies the type being aliased by name.
+
+``array``
+  An LLVM array. There are two additional fields:
+
+  - ``size``: An integer giving the length of the array.
+  - ``element type``: An :ref:`LLVM type<llvm-types>` describing the array elements.
+
+``function``
+  A function type. There are three additional fields:
+
+  - ``return type``: An :ref:`LLVM type<llvm-types>` describing the return type of the function.
+  - ``argument types``: A list of :ref:`LLVM types<llvm-types>` describing the arguments of the function.
+  - ``varargs``: A Boolean indicating whether this function takes a variable number of arguments.
+
+``pointer``
+  A pointer type. There is an additional field ``to type``, an :ref:`LLVM type<llvm-types>` describing the
+  referent type of the pointer.
+
+``struct``
+  A structure type. There is an additional field ``fields``, a list of :ref:`LLVM types<llvm-types>` describing
+  the structure fields.
+
+``packed struct``
+  A packed structure type. There is an additional field ``fields``, a list of :ref:`LLVM types<llvm-types>` describing
+  the structure fields.
+
+``vector``
+  An LLVM vector. There are two additional fields:
+
+  - ``size``: An integer giving the length of the array.
+  - ``element type``: An :ref:`LLVM type<llvm-types>` describing the array elements.
+
+``opaque``
+  An opaque structure type.
+
+.. _jvm-types:
+
+JVM Types
+=========
+
+As with LLVM types, there is a tag field named ``type``. This tag value can be:
+
+``primitive type``
+  A JVM primitive type. There is an additional field ``primitive`` which can be any of the following:
+
+  - ``boolean``: A JVM Boolean.
+  - ``byte``: A JVM byte.
+  - ``char``: A JVM character.
+  - ``double``: A JVM double-precision float.
+  - ``float``: A JVM single-precsion float.
+  - ``int``: A JVM integer.
+  - ``long``: A JVM long integer.
+  - ``short``: A JVM short integer.
+
+``array type``
+  A JVM array. There are two additional fields:
+
+  - ``size``: An integer giving the length of the array.
+  - ``element type``: An :ref:`JVM type<jvm-types>` describing the array elements.
+
+``class type``
+  A JVM class. There is an additional field ``class name`` which identifies the class.
```

Comment:
We should specify how the class name is encoded (dots or slashes), though I'm not sure the underlying code is finished enough to have solidified that decision.

Reply:
From some quick spelunking, I _think_ class names are currently encoded using dots. I can update these docs to reflect that, if you'd like.

Reply:
Sure. That sounds great.